### PR TITLE
Advanced solution

### PR DESCRIPTION
**Pull request by skviropitek**

### CODE REVIEW

**Review by kerrrusha — CHANGES_REQUESTED** [review]
your checks failed, please fix them first
image

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class ColorSupplier {
+    static final String[] Colors = {"Red", "Blue", "Green", "Yellow", "Purple", "Orange"};
```
Comment (MentorMA) [inline src/main/java/core/basesyntax/ColorSupplier.java:6]: It should be enums

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+    public static String getRandomColor() {
```
Comment (MentorMA) [inline src/main/java/core/basesyntax/ColorSupplier.java:8]: Why it static?

File: src/main/java/core/basesyntax/Figure.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    public String getColor() {
+        return color;
+    }
+
+    public void setColor(String color) {
+        this.color = color;
+    }
+
+    abstract double getArea();
```
Comment (MentorMA) [inline src/main/java/core/basesyntax/Figure.java:18]: Create interface AreaCalculator with this method

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+                double trapezoidHeight = random.nextDouble() * 10 + 1;
+                return new IsoscelesTrapezoid(color, trapezoidBase1, trapezoidBase2,
+                    trapezoidHeight);
+            default:
+                return null;
+        }
+    }
+
+    public static Figure getDefaultFigure() {
+        return new Circle("White", 10);
```
Comment (MentorMA) [inline src/main/java/core/basesyntax/FigureSupplier.java:37]: It should ve constants

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                return new RightTriangle(color, triangleFirstLeg, triangleSecondLeg);
+            case 3:
+                double circleRadius = random.nextDouble() * 10 + 1;
+                return new Circle(color, circleRadius);
+            case 4:
+                double trapezoidBase1 = random.nextDouble() * 10 + 1;
+                double trapezoidBase2 = random.nextDouble() * 10 + 1;
+                double trapezoidHeight = random.nextDouble() * 10 + 1;
+                return new IsoscelesTrapezoid(color, trapezoidBase1, trapezoidBase2,
+                    trapezoidHeight);
```
Comment (MentorMA) [inline src/main/java/core/basesyntax/FigureSupplier.java:30]: magic numbers, write private methods for creating every figure

File: src/main/java/core/basesyntax/Main.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+        for (int i = 0; i < 5; i++) {
+            figures[i] = FigureSupplier.getRandomFigure();
+        }
+        for (int i = 5; i < 10; i++) {
+            figures[i] = FigureSupplier.getDefaultFigure();
+        }
+        for (Figure figure : figures) {
+            figure.draw();
+        }
```
Comment (MentorMA) [inline src/main/java/core/basesyntax/Main.java:15]: you don`t need too many loops

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ColorSupplier.java (hunk 2 above)
Comment (Vasul-Pankiv) [reply src/main/java/core/basesyntax/ColorSupplier.java:8]: make this method non static

File: src/main/java/core/basesyntax/FigureSupplier.java (hunk 5 above)
Comment (Vasul-Pankiv) [reply src/main/java/core/basesyntax/FigureSupplier.java:30]: not fixed

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]
do not ignore comments, if you have questions - ask it in chat

File: src/main/java/core/basesyntax/Circle.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        return Math.PI * radius * radius;
+    }
+
+    @Override
+    public void draw() {
+        System.out.println("Circle - Color: " + getColor()
+                + ", Radius: " + radius
+                + ", Area: " + area());;
+    }
+
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/Circle.java:30]:
```suggestion
```

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class ColorSupplier {
+    private static final Random RANDOM = new Random();
+
+    public String getRandomColor() {
+        Color[] colors = Color.values();
+        return colors[RANDOM.nextInt(colors.length)].toString();
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/ColorSupplier.java:10]:
```suggestion
        return colors[RANDOM.nextInt(colors.length)].name();
```

File: src/main/java/core/basesyntax/Figure.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package core.basesyntax;
+
+public abstract class Figure implements Drawable, AreaCalculator {
+    public static final String default_color = "White";
+    public static final int default_radius = 10;
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/Figure.java:5]: please follow a naming convention for constants

File: src/main/java/core/basesyntax/Figure.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    private String color;
+    private double area;
+
+    public Figure(String color) {
+        this.color = color;
+    }
+
+    public static Figure getDefaultFigure() {
+        return new Circle(default_color, default_radius);
+    }
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/Figure.java:16]: do not add this method to this class

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            case 1:
+                return createRectangle(color);
+            case 2:
+                return createRightTriangle(color);
+            case 3:
+                return createCircle(color);
+            case 4:
+                return createIsoscelesTrapezoid(color);
+            default:
+                return null;
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/FigureSupplier.java:31]:
```suggestion
            default:
                return createIsoscelesTrapezoid(color);
```

File: src/main/java/core/basesyntax/Square.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        return sideLength * sideLength;
+    }
+
+    @Override
+    public void draw() {
+        System.out.println("Square color: " + getColor()
+                + ", Side length = " + sideLength
+                + "Area = " + area());
+    }
+
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/Square.java:30]:
```suggestion
```

File: src/main/java/core/basesyntax/RightTriangle.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    public void draw() {
+        System.out.println("Right triangle color: " + getColor()
+                + ", First leg length = " + firstLeg
+                + ", Second leg length = " + secondLeg
+                + "Area = " + area());
+    }
+
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/RightTriangle.java:41]:
```suggestion
```

File: src/main/java/core/basesyntax/Rectangle.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        return length * width;
+    }
+
+    @Override
+    public void draw() {
+        System.out.println("Rectangle color: " + getColor()
+                + ", Length = " + length + ", Width = " + width
+                + "Area = " + area());
+    }
+
```
Comment (sarakhmen) [inline src/main/java/core/basesyntax/Rectangle.java:40]:
```suggestion
```

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Let's improve your solution :)

File: src/main/java/core/basesyntax/Color.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax;
+
+public enum Color {
+    RED,
+    BLUE,
+    GREEN,
+    PURPLE,
+    YELLOW,
+    ORANGE
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Color.java:9]:
```suggestion
    ORANGE,
    WHITE
```

File: src/main/java/core/basesyntax/Figure.java [hunk 16]
```diff
@@ -0,0 +1,29 @@
+package core.basesyntax;
+
+public abstract class Figure implements Drawable, AreaCalculator {
+    public static final String DEFAULT_COLOR = "White";
+    public static final int DEFAULT_RADIUS = 10;
+
+    private String color;
+    private double area;
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Figure.java:8]:
```suggestion
```

File: src/main/java/core/basesyntax/Figure.java [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+    public String getColor() {
+        return color;
+    }
+
+    public void setColor(String color) {
+        this.color = color;
+    }
+
+    public double getArea() {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Figure.java:22]: remove get and set for area field

File: src/main/java/core/basesyntax/Figure.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+    public Figure(String color) {
+        this.color = color;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public void setColor(String color) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Figure.java:18]: remove, you set color by using constractor

File: src/main/java/core/basesyntax/Figure.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
+package core.basesyntax;
+
+public abstract class Figure implements Drawable, AreaCalculator {
+    public static final String DEFAULT_COLOR = "White";
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Figure.java:4]: Replace DEFAULT_COLOR and DEFAULT_RADIUS fo FigureSupplier

Also, add WHITE to Color enum and use it for constant DEFAULT_COLOR

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.util.Random;
+
+public class FigureSupplier {
+    private static final int FIGURE_TYPES = 5;
+    private static final double DIMENSION_RANGE = 10.0;
+    private final Random random;
+    private final ColorSupplier colorSupplier;
+
+    public FigureSupplier(Random random, ColorSupplier colorSupplier) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/FigureSupplier.java:14]: let's make FigureSupplier more isolated, so initialise these parameters during the declaration
```
private final Random random = new Random();
private final ColorSupplier colorSupplier = new ColorSupplier();
```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        double radius = random.nextDouble() * DIMENSION_RANGE + 1;
+        return new Circle(color, radius);
+    }
+
+    private Figure createIsoscelesTrapezoid(String color) {
+        double base1 = random.nextDouble() * DIMENSION_RANGE + 1;
+        double base2 = random.nextDouble() * DIMENSION_RANGE + 1;
+        double height = random.nextDouble() * DIMENSION_RANGE + 1;
+        return new IsoscelesTrapezoid(color, base1, base2, height);
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/FigureSupplier.java:67]: Create separate method for `random.nextDouble() * DIMENSION_RANGE + 1`;

File: src/main/java/core/basesyntax/Main.java [hunk 22]
```diff
@@ -0,0 +1,28 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class Main {
+    public static void main(String[] args) {
+
+        Random random = new Random();
+        ColorSupplier colorSupplier = new ColorSupplier();
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Main.java:9]: Remove these variables, see the comment above

File: src/main/java/core/basesyntax/Main.java [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+
+        FigureSupplier figureSupplier = new FigureSupplier(random, colorSupplier);
+
+        Figure[] figures = new Figure[10];
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Main.java:13]: Create constant for 10 is a magic number

File: src/main/java/core/basesyntax/Main.java [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+        for (int i = 0; i < 10; i++) {
+            if (i < 5) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Main.java:15]:
```suggestion
            if (i < SIZE_OF_ARRAY / 2) {
```

File: src/main/java/core/basesyntax/Main.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+                figures[i] = figureSupplier.getRandomFigure();
+            } else {
+                figures[i] = figureSupplier.getDefaultFigure();
+            }
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Main.java:19]:
```suggestion
            }
            figures[i].draw();
```

File: src/main/java/core/basesyntax/Main.java [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+        for (Figure figure : figures) {
```
Comment (Elena-Bruyako) [inline src/main/java/core/basesyntax/Main.java:22]: remove this loop

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Pull request closed without merging**